### PR TITLE
feature: replace bespoke regex parsing with Lark

### PR DESCRIPTION
**Pull request by Jaguar0625**
part 1:
- define initial grammar
- add AST object model
- support legacy YAML generation from AST objects

### CODE REVIEW

File: catparser/CatsLarkParser.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+        # region enum
+
+        @staticmethod
+        def enum(tokens):
+            return Enum(tokens)
+
+        @staticmethod
+        def enum_child(tokens):
+            return CatbufferTransformer.statement(tokens)
```
Comment (gimre-xymcity) [inline catparser/CatsLarkParser.py:76]: hmm, dumb question, but why is this needed?

**Review by gimre-xymcity — COMMENTED** [review]
reviewed everything ex tests

File: catparser/grammar/catbuffer.lark [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+fill_based_array: "__FILL__"
+
+conditional_expression: "if" (_dec_or_hex_number | CONST_PROPERTY_NAME) CONDITIONAL_OPERATION PROPERTY_NAME
+CONDITIONAL_OPERATION: "not equals" | "equals" | "in" | "not in"
+
+_integer_or_enum_const: (FIXED_SIZE_INTEGER "," _dec_or_hex_number | USER_TYPE_NAME "," CONST_PROPERTY_NAME)
+
+_dec_or_hex_number: DEC_NUMBER | HEX_NUMBER
+DEC_NUMBER: DIGIT+
+HEX_NUMBER: "0x" ("A".."F" | DIGIT)+
```
[the remark this answers is not in the document]
Comment (Jaguar0625) [reply catparser/grammar/catbuffer.lark:48]: i wanted to require upper case 😅

File: catparser/grammar/catbuffer.lark [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+struct_child: [comment] (struct_field | struct_field_const | struct_field_reserved | struct_field_inline | struct_inline)
+struct_field: (VALUE_PROPERTY_NAME_PLACEHOLDER | PROPERTY_NAME) "=" (USER_TYPE_NAME | FIXED_SIZE_INTEGER | array_expression) [conditional_expression] _NL
+struct_field_const: CONST_PROPERTY_NAME "=" "make_const" "(" _integer_or_enum_const ")" _NL
+struct_field_reserved: PROPERTY_NAME "=" "make_reserved" "(" _integer_or_enum_const ")" _NL
+struct_field_inline: PROPERTY_NAME "=" "inline" USER_TYPE_NAME _NL
+struct_inline: "inline" USER_TYPE_NAME _NL
+
+array_expression: "array" "(" (FIXED_SIZE_INTEGER | USER_TYPE_NAME) "," _array_size_expression ")"
+_array_size_expression: count_based_array | size_based_array | fill_based_array
+count_based_array: PROPERTY_NAME ["," "sort_key" "=" PROPERTY_NAME] | _dec_or_hex_number
```
[the remark this answers is not in the document]
Comment (Jaguar0625) [reply catparser/grammar/catbuffer.lark:37]: some of our state objects, with fixed size arrays e.g.
```
buckets = array(HeightActivityBucket, 5)
```

File: catparser/grammar/catbuffer.lark [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+struct: ["inline"] "struct" USER_TYPE_NAME _NL _INDENT struct_child* _DEDENT
+struct_child: [comment] (struct_field | struct_field_const | struct_field_reserved | struct_field_inline | struct_inline)
+struct_field: (VALUE_PROPERTY_NAME_PLACEHOLDER | PROPERTY_NAME) "=" (USER_TYPE_NAME | FIXED_SIZE_INTEGER | array_expression) [conditional_expression] _NL
+struct_field_const: CONST_PROPERTY_NAME "=" "make_const" "(" _integer_or_enum_const ")" _NL
+struct_field_reserved: PROPERTY_NAME "=" "make_reserved" "(" _integer_or_enum_const ")" _NL
+struct_field_inline: PROPERTY_NAME "=" "inline" USER_TYPE_NAME _NL
+struct_inline: "inline" USER_TYPE_NAME _NL
+
+array_expression: "array" "(" (FIXED_SIZE_INTEGER | USER_TYPE_NAME) "," _array_size_expression ")"
+_array_size_expression: count_based_array | size_based_array | fill_based_array
```
[the remark this answers is not in the document]
Comment (Jaguar0625) [reply catparser/grammar/catbuffer.lark:36]: size is a byte count (instead of an element count) and currently only used in Aggregate
```
	transactions = array(EmbeddedTransaction, @size=payload_size)
```

i don't think the name is great either, any suggestions?

File: catparser/CatsLarkParser.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        # region comment
+
+        @staticmethod
+        def comment(tokens):
+            return Comment(tokens[0])
+
+        @staticmethod
+        def statement(tokens):
+            tokens[1].comment = tokens[0]
+            return tokens[1]
```
[the remark this answers is not in the document]
Comment (Jaguar0625) [reply catparser/CatsLarkParser.py:56]: i couldn't find any way to do that

File: catparser/CatsLarkParser.py (hunk 1 above)
Comment (Jaguar0625) [reply catparser/CatsLarkParser.py:76]: so we can attach comments to enum value definitions.
technically we could merge enum_child and enum_value, but i wanted to keep semantics similar to struct.

File: catparser/grammar/catbuffer.lark (hunk 4 above)
[the remark this answers is not in the document]
Comment (gimre-xymcity) [reply catparser/grammar/catbuffer.lark:36]: was thinking of `variable_elements_array`, but not great as well

File: catparser/ast.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
+from abc import ABC, abstractmethod
+
+from lark import Token
+
+
+def _get_token_value(token):
```
[the remark this answers is not in the document]
Comment (gimre-xymcity) [reply catparser/ast.py:6]: no opinion on this one (`get_token_value` is clear for me)

File: tests/test_ast.py [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+class EnumTests(unittest.TestCase):
+    def _test_can_create_enum(self, comment, expected_comment_descriptor):
+        # Act:
+        model = Enum(['ColorShade', FixedSizeInteger('uint32'), EnumValue(['RED', 0xFF0000]), EnumValue(['GREEN', 0x00FF00])])
+        model.comment = comment
+
+        # Assert:
+        self.assertEqual('ColorShade', model.name)
+        self.assertEqual('uint32', model.base.short_name)
+        self.assertEqual([('RED', 0xFF0000), ('GREEN', 0x00FF00)], [(value.name, value.value) for value in model.values])
```
Comment (gimre-xymcity) [inline tests/test_ast.py:149]: for enum_value in ? (no strong opinion)

**Review by gimre-xymcity — APPROVED** [review]

File: catparser/grammar/catbuffer.lark (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jaguar0625) [reply catparser/grammar/catbuffer.lark:36]: will leave original name for now
